### PR TITLE
DTLS memory optimization when used with SCTP as transport

### PR DESCRIPTION
**Pull request by NaveenShivanna86**
A DTLS connection memory can be optimized by freeing the `init_buf` after the handshake is completed when the transport is SCTP . This `init_buf` is required to handle the re-transmitted messages like 'FINISHED' from client. `init_buf` memory is around 22KB in case of DTLS.
DTLS over UDP can still get the re-transmitted message but there is no re-transmission in case of DTLS over SCTP.

RFC 6083:
  SCTP provides a reliable and in-sequence transport service for DTLS
   messages that require it.  As per section 4.4.  Therefore, DTLS
   procedures for re-transmissions MUST NOT be used.

Hence `init_buf` can be freed when DTLS over SCTP as transport is used.

### CODE REVIEW

**Comment by NaveenShivanna86** [comment]
Requesting to review.

**Comment by mattcaswell** [comment]
We will need a CLA for this. Please see:

https://www.openssl.org/policies/cla.html

**Comment by NaveenShivanna86** [comment]
About CLA, I have sent the signed scan copy to legal@opensslfoundation.org on 26-08-2019.

**Comment by mattcaswell** [comment]
Close/reopen to kick CLA bot.

**Comment by NaveenShivanna86** [comment]
Requesting to review.

**Comment by NaveenShivanna86** [comment]
Requesting to review.

**Comment by mattcaswell** [comment]
Ping @openssl/committers

**Comment by NaveenShivanna86** [comment]
Hi, Requesting to review.

**Comment by mattcaswell** [comment]
@openssl/committers - this one should be easy to review.

File: ssl/statem/statem_lib.c [hunk 1]
```diff
@@ -1039,14 +1039,25 @@ WORK_STATE tls_finish_handshake(SSL *s, WORK_STATE wst, int clearbufs, int stop)
...
     if (clearbufs) {
-        if (!SSL_IS_DTLS(s)) {
+        if (!SSL_IS_DTLS(s)
+#ifndef OPENSSL_NO_SCTP
+                /*
+		 * RFC6083: SCTP provides a reliable and in-sequence transport service for DTLS
+                 * messages that require it. Therefore, DTLS procedures for retransmissions
+		 * MUST NOT be used.
+                 * Hence the init_buf can be cleared when DTLS over SCTP as transport is used.
+                 */
```
Comment (mspncp) [inline ssl/statem/statem_lib.c]: Looks like there is mixed whitespace in lines 1044-1049?

File: ssl/statem/statem_lib.c [hunk 2]
```diff
@@ -1039,14 +1039,25 @@ WORK_STATE tls_finish_handshake(SSL *s, WORK_STATE wst, int clearbufs, int stop)
...
+	        || BIO_dgram_is_sctp(SSL_get_wbio(s))
+#endif
+	        ) {
             /*
-             * We don't do this in DTLS because we may still need the init_buf
+             * We don't do this in DTLS over UDP because we may still need the init_buf
```
Comment (mspncp) [inline ssl/statem/statem_lib.c]: Just a curious question maybe: Why did you add "over UDP"? I thought Datagram TLS was synonymous to TLS over UDP? Is this just a hendiadys, or are there other DTLS scenarios?

Comment (mspncp) [reply ssl/statem/statem_lib.c]: Oh, forget my stupid question. I should have googled SCTP first ;-)

File: ssl/statem/statem_lib.c (hunk 1 above)
Comment (mspncp) [reply ssl/statem/statem_lib.c]: I would like to see this description (possibly in an abbreviated form) in the body of the commit message, too.

**Comment by mspncp** [comment]
Note: Travis failure is unrelated (apt-get failed): https://travis-ci.org/openssl/openssl/jobs/577372913

File: ssl/statem/statem_lib.c (hunk 1 above)
Comment (NaveenShivanna86) [reply ssl/statem/statem_lib.c]: I have worked on your comments, please review.

File: ssl/statem/statem_lib.c [hunk 3]
```diff
@@ -1039,14 +1039,25 @@ WORK_STATE tls_finish_handshake(SSL *s, WORK_STATE wst, int clearbufs, int stop)
...
     if (clearbufs) {
-        if (!SSL_IS_DTLS(s)) {
+        if (!SSL_IS_DTLS(s)
+#ifndef OPENSSL_NO_SCTP
+                /*
+                 * RFC6083: SCTP provides a reliable and in-sequence transport service for DTLS
+                 * messages that require it. Therefore, DTLS procedures for retransmissions
+                 * MUST NOT be used.
+                 * Hence the init_buf can be cleared when DTLS over SCTP as transport is used.
+                 */
```
Comment (mspncp) [inline ssl/statem/statem_lib.c]: The comment block is indented too far by four bytes. But never mind, I will fix it when I merge your pull request.

**Comment by mspncp** [comment]
Same question here @mattcaswell: does your approval hold for 1.1.1, too?

**Comment by mattcaswell** [comment]
Yes - approval for 1.1.1 too.

**Comment by mspncp** [comment]
Merged to master and 1.1.1. Thank you @NaveenShivanna86!

e7c27a6c37 'init_buf' memory can be freed when DTLS is used over SCTP (not over UDP). [master]
a80de34b27 'init_buf' memory can be freed when DTLS is used over SCTP (not over UDP). [1.1.1]
